src/hermes/model/types/__init__.py

```diff
@@ -21,56 +21,48 @@
             "ld_container": lambda c, **_: c,
             "json": lambda c, **_: c.compact(),
             "expanded_json": lambda c, **_: c.ld_value,
-        },
+        }
     ),
 
     # Wrap item from ld_dict in ld_list
-    (ld_list.is_ld_list, dict(ld_container=ld_list)),
-    (
-     lambda c: isinstance(c, list) and all(isinstance(item, dict) for item in c),
-     dict(ld_container=lambda c, **kw: ld_list([{"@list": c}], **kw))
-    ),
+    (ld_list.is_ld_list, {"ld_container": ld_list}),
+    (lambda c: isinstance(c, list), {"ld_container": lambda c, **kw: ld_list([{"@list": c}], **kw)}),
 
     # pythonize items from lists (expanded set is already handled above)
-    (ld_container.is_json_id, dict(python=lambda c, **_: c["@id"])),
-    (ld_container.is_typed_json_value, dict(python=ld_container.typed_ld_to_py)),
-    (ld_container.is_json_value, dict(python=lambda c, **_: c["@value"])),
-    (ld_list.is_container, dict(ld_container=lambda c, **kw: ld_list([c], **kw))),
-    (ld_dict.is_json_dict, dict(ld_container=lambda c, **kw: ld_dict([c], **kw))),
+    (ld_container.is_json_id, {"python": lambda c, **_: c["@id"]}),
+    (ld_container.is_typed_json_value, {"python": lambda c, **kw: ld_container.typed_ld_to_py([c], **kw)}),
+    (ld_container.is_json_value, {"python": lambda c, **_: c["@value"]}),
+    (ld_list.is_container, {"ld_container": lambda c, **kw: ld_list([c], **kw)}),
+    (ld_dict.is_json_dict, {"ld_container": lambda c, **kw: ld_dict([c], **kw)}),
+    (lambda v: isinstance(v, str), {"python": lambda v, parent, **_: parent.ld_proc.compact_iri(parent.active_ctx, v)}),
 
     # Convert internal data types to expanded_json
-    (lambda c: ld_container.is_json_id(c) or ld_container.is_json_value(c), dict(expanded_json=lambda c, **_: [c])),
-    (ld_dict.is_json_dict, dict(expanded_json=lambda c, **kw: ld_dict.from_dict(c, **kw).ld_value)),
-    (ld_dict.is_ld_dict, dict(expanded_json=lambda c, **kw: ld_dict.from_dict(c[0], **kw).ld_value)),
+    (lambda c: ld_container.is_json_id(c), {"expanded_json": lambda c, **_: c}),
+    (lambda c: ld_container.is_ld_id(c), {"expanded_json": lambda c, **_: c[0]}),
+    (lambda c: ld_container.is_json_value(c), {"expanded_json": lambda c, **_: [c]}),
+    (lambda c: ld_container.is_ld_value(c), {"expanded_json": lambda c, **_: c}),
```

Comment:
Why is the callback wrapped in an additional `lambda c` here?

However, this seems like a reasonable and important change... maybe an additinal hint in the comment about the side effects of the ordering of especially these rules might be helpful. (I'll also have look into the commit history to better understand the reasons behind this).